rclcpp/src/rclcpp/experimental/timers_manager.cpp

```diff
@@ -191,6 +192,7 @@ std::chrono::nanoseconds TimersManager::get_head_timeout_unsafe()
     }
     head_timer = locked_heap.front();
   }
+  head_was_cancelled = head_timer->is_canceled();
```

Comment:
do we need an extra variable? what does `time_until_trigger` returns when the timer is cancelled?

Reply:
Yes, I found it was necessary. `time_until_trigger` actually returns `std::chrono::nanoseconds::max()` when the timer is cancelled, which I think is how the bug manifests -- that then forces the call to `timers_cv_.wait(...)`

Reply:
It looks like the `get_head_timeout_unsafe` is essentially trying to indicate 3 different scenarios:
 - it should indicate when the heap is empty, and it currently does that by returning `std::chrono::nanoseconds::max()`
 - it should indicate when the front timer in the heap is cancelled, and it currently does that by setting the `head_was_cancelled` bool variable
 - if the front timer is good, then it should indicate the time until trigger

I think that this behavior is not very intuitive.

Moreover, what happens if there's only 1 timer which is cancelled? Will the heap keep polling it?

Reply:
Great point! Refactored to use `std::optional` as the return value.